src/textual/widget.py

```diff
@@ -105,6 +107,25 @@ def __init__(
     show_vertical_scrollbar = Reactive(False, layout=True)
     show_horizontal_scrollbar = Reactive(False, layout=True)
 
+    def get_box_model(self, container_size: Size, parent_size: Size) -> BoxModel:
+        box_model = get_box_model(
+            self.styles,
+            container_size,
+            parent_size,
+            self.get_content_width,
+            self.get_content_height,
+        )
+        return box_model
+
+    def get_content_width(self, container_size: Size, parent_size: Size) -> int:
+        console = self.app.console
+        renderable = self.render()
+        measurement = Measurement.get(console, console.options, renderable)
+        return measurement.maximum
+
+    def get_content_height(self, container_size: Size, parent_size: Size) -> int:
+        return container_size.height
```

Comment:
Rich has a way of measuring the renderable via `__rich_measure__` which is what `get_content_width` uses. Most renderables can report their width fairly cheaply.

Getting the height would require rendering it, and it would need to know what width to use before it could report the height. So it is an altogether more complex calculation.

I've made the auto height return the height of the container for now. I would like to make it genuinely fit the content, but I'm going to have to think about how to do that elegantly.